In-progress draft of Resource metadata (FOR COLLABORATIVE EDITING) (#32)
solid/data-interoperability-panel: resource-metadata/hypothesis.md

# Resource Metadata Model

## Background

This document introduces a mechanism for linking to metadata about resources in the Solid Ecosystem using HTTP Link headers. Examples of this linking mechanism include:

> - A binary JPEG image with a Link header, pointing to a description that has an RDF representation.

Reviewer: I have created dedicated issue for media resources and multiple encodings of same creative work #33 . Here I would like to clarify if we talk about description of each of those encodings or description of the creative work which can have numerous encodings?

Reviewer: If we just talk about different representation of the same resource, I think we could just use `rel=alternate` together with `type=image/jpeg` http://microformats.org/wiki/rel-alternate#With_type

> When used in combination with the `type` attribute (with a value other than that of the document itself; e.g. other than `"text/html"`), `rel="alternate"` means a link to a representation of the contents of the current document in a different format, as designated by the `type` attribute. 

If we talk about relating a creative work to multiple different encodings, each of which can have their own encoding specific rdf representation, i think should discuss it in #33

Author: Would you say that the proposed approach we describe here gives us the basic capabilities needed to make these associations (for example in the content of the resource description metadata resource)?

For any defined metadata type available for a given resource, all representations of that resource MUST include a Link header pointing to the location of each metadata resource. For example, as defined by the Solid Web Access Control specification, a client can use this mechanism to discover the location of an ACL resource:

Link: <https://example.com/resource?acl>; rel="acl"

### Discovery of Annotated Resource

Certain metadata resource types MAY require the server to include a link back to the annotated resource using an appropriate link relation. For example, LDP defines a bidirectional discovery mechanism for RDF descriptions of NonRDF resources, via Link headers:

Link: <https://example.com/binary?description>; rel="description"

along with:

Link: <https://example.com/binary>; rel="describedby"

For cases where link relations are not defined by IANA, a URL can be used. For example:

Link: <https://example.com/resource?meta>;
        rel="https://example.org/ns#hasMetadata"

and:

Link: <https://example.com/resource>;
        rel="https://example.org/ns#isMetadataOf"

> String-based link relations, such as in the examples above, must be registered with IANA. But it is also possible to use custom relation types by using a full IRI. The Linked Data Platform, the Linked Data Notification and the Web Annotation specifications make use of full IRIs in the rel attribute.

Reviewer 2: ```suggestion
String-based link relations, such as in the examples above, must be registered with IANA. But it is also possible to use custom relation types by using a full IRI. The [Linked Data Platform](https://www.w3.org/TR/ldp/), the [Linked Data Notifications](https://www.w3.org/TR/ldn/), and the [Web Annotation Protocol](http://www.w3.org/TR/annotation-protocol/) specifications make use of full IRIs in the `rel` attribute.
```

https://github.com/solid/specification/issues/58
https://github.com/solid/specification/issues/42
https://github.com/solid/specification/issues/31
https://github.com/solid/specification/issues/131

ACL resources as defined by [Web Access Control](https://github.com/solid/web-access-control-spec) MUST be supported.

MUST be deleted when resource is deleted.

Authorization Model:

Discovered via ```rel=acl``` and ```rel=http://www.w3.org/ns/solid/terms#resource```

### Resource Description

Resource description provides a general mechanism for resource annotation, such as for providing descriptive metadata for a binary file. MUST be supported. MUST be deleted when resource is deleted.

Authorization Model:

Discovered via ```rel=describedby``` and ```rel=describes```

### Shape Constraint

Enforces conformance of the resource to an associated SHACL or ShEx data shape. MUST be supported. MUST be deleted when resource is deleted.

Authorization Model:

Discovered via ```rel=http://www.w3.org/ns/solid/terms#shape``` and ```rel=http://www.w3.org/ns/solid/terms#resource```

### Server Managed

Server-managed metadata not otherwise included in HTTP headers. MUST be supported.

> __Note: This may be too broad to be able to properly determine what the authorization rule, or lifecycle should be. May need to consider breaking this up into specific types.__

Reviewer: Since we don't expose them over HTTP I don't think authorization rules apply to them. We may just need to define something to make them portable.

Author: > I've realized that I don't understand how it currently works, and I still miss some information to understand it.
> 
> The main thing I fail to understand relates to the question: "what is a resource in its own right?"
> 
> I've observed that if you add triples to a container, those triples will be added to `.meta`. In that case, `.meta` is not a resource in its own right, I suppose. You wouldn't interact with it, you'd interact with the container.
> 
> Then, I suppose that if you have an image, and you wish to add metadata to the image, you would do that to a separate resource, and you would get metadata by dereferencing a separate resource. So, in that case, the metadata resource is a resource in its own right. Concievably, you could discriminate between the representations by media type though.
> 
> It would be nice to have clarification around that.

Take a look at latest changes, which stipulate that metadata resources on the same server follow the LDPR interaction model, and then (in implementation patterns) that servers know/maintain which resources are metadata, and can apply special handling to them.

Author: Per the discussion in the panel meeting yesterday this was poorly worded and a bit misleading as written because it was incomplete. The next round will complete the section and address this so will resolve for now.

Discovered via ```rel=http://www.w3.org/ns/solid/terms#metadata``` and ```rel=http://www.w3.org/ns/solid/terms#resource```

### Configuration

Configuration metadata, e.g. settings for memento creation. MUST be supported. MUST be deleted when resource is deleted.

Authorization Model:

Discovered via ```rel=http://www.w3.org/ns/solid/terms#configuration```

## Non-Reserved Types

A server may support other metadata types.

## Implementation Patterns

There are many ways a server could implement these features. A file-based server could have a special naming scheme reserved for these metadata resources. Alternatively, a server could represent every resource internally as a dataset, storing each separate type of metadata in its own named graph.

[ Include example of file based ]

[ Include example of graph based ]

> ## Open Questions

Reviewer: I think we should create dedicated issue of each of those questions.

# Resource Metadata Model

> ## Background

Reviewer 3: High level suggestion: What people often conflate is metadata that's loosely coupled with a resource for the purpose of extending interaction possibilities or certain behaviours that comes out of it from the kind of data that extends the description of a resource. Making that distinction up front may help. The classic example has to do with how to deal with providing a human-readable label for a container (or from LDP, updating "minimal-container triples"). That use case or anything that's about self-describing resources is outside the scope of "resource metadata" (as I see it and hopefully I'm not off the mark with what the panel intends to do). What's described here is really about the ones that's highlighted: access control, annotation, shapes, server-managed, configuration, and possibly others defined elsewhere. This is important because we still want the primary resource to be useful in its own right without forcing the client to navigate to the the related meta resource (eg. just to get a hold of container's label) - that's besides annotating non-RDF Sources. That also opens up the possibility to progressive enhance so to speak, and if so desired.

It may help to just change the order of information in the Background.

I sense some redundancy with the two example blocks. Perhaps that can be a single list that's tied to use cases eg. at least one for each metadata type.

The section should be condensed as much as possible. It could just be an "Overview" - let's see..

Author: Incorporated suggestions and condensed overview.

*This section is non-normative*

This document introduces a mechanism for linking to metadata about resources in the Solid Ecosystem using HTTP Link headers. Examples of this linking mechanism include:

- A binary JPEG image with a Link header, pointing to a description that has an RDF representation.

> - An LDP container with a Link header, pointing to an ACL resource that governs access controls for the contained resources.

Reviewer 3: I'd suggest not to mention LDP container but just go with "container" because the Solid container (`/`) is not strictly defined as an LDP container. So, just `s/An LDP/A` would suffice.

Author: Removed LDP

A given resource might link to zero or more such related metadata representations. Having a unified model to describe how clients and Solid servers interact with and process these resources will help clarify expectations, while also providing a shared pattern for extending the features of Solid server implementations.

The metadata model described in this document makes it possible for clients to decorate resources with structured descriptive data. Those metadata may serve as supplementary descriptions or, when supported by a Solid server implementation, they may influence the behavior of the resources.

Some examples include:

- An ACL resource controls how a Solid server makes authorization decisions for a container and any child resources.
- A shape constraint resource may limit the RDF structures that can be added to a resource or container.
- A “configuration” resource may control how a resource is versioned, or which indexes are exposed for it.

Access to different categories of metadata may require different levels of privilege, which must be specified as part of the definition for a given metadata type.

> ## Metadata Discovery

Reviewer 3: To give an overview, I think text and table along the the following lines can help (with some inspiration from the fragment specification table under https://www.w3.org/TR/annotation-model/#model-18 ):

"
The following are some of the resource metadata types and the relation URIs that are used to discover the resource metadata. Other metadata types and relations may also be used.

Type | Relation | Description
--- | --- | ---
Access control | `acl` or `http://www.w3.org/ns/solid/terms#acl` | [[WAC](http://solid.github.io/web-access-control-spec/)]
Resource description  | `describedby` or `https://www.w3.org/ns/iana/link-relations/relation#describedby` | ...
Shape | `http://www.w3.org/ns/solid/terms#shape` | ...
Server managed | `http://www.w3.org/ns/solid/terms#servermanaged` | ...
Configuration | `http://www.w3.org/ns/solid/terms#configuration` | ...

"

> Given the URL of a resource, a client can discover the metadata resources by issuing a `GET` (or `HEAD`) request and inspecting the `Link` headers in the response. The [`rel={attribute}`](https://tools.ietf.org/html/rfc8288) will define the relationship to the target URL.

Reviewer 3: Is it restricted to `GET` and `HEAD`? How about `OPTIONS`?

Should `rel={attribute}` be `rel={relationtype}` ?

Author: Added OPTIONS and changed to ```rel={relation-type}```

For any defined metadata type available for a given resource, all representations of that resource MUST include a Link header pointing to the location of each metadata resource. For example, as defined by the Solid [Web Access Control specification](https://github.com/solid/web-access-control-spec), a client can use this mechanism to discover the location of an ACL resource:

> Link: <https://example.com/resource?acl>; rel="acl"

Reviewer 3: I acknowledge the intention with the example use of `?acl` but to minimise any confusion eg. is the ACL resource somehow a query on a resource (or even if a Solid server is capable of taking query params), as well as to emphasise the notion that the ACL resource is really a distinct resource, use something like `https://example.org/resource.acl`

Metadata discovered through a Link header for a given resource is considered to be *directly associated* with that resource.

> ### Discovery of Annotated Resource

Reviewer 3: Consider information table like in https://github.com/solid/data-interoperability-panel/pull/32/files#r371016439 or have a single table combining relations from both sides.

> Certain metadata resource types MAY require the Solid server to include a link back to the annotated resource using an appropriate link relation. For example, LDP defines a bidirectional discovery mechanism for RDF descriptions of NonRDF resources, via Link headers:

Reviewer 3: "MAY require"?

No need to mention LDP behaviour.

`s/NonRDF/non-RDF`

Author: Adjusted language for clarity.

> Link: <https://example.com/binary?description>; rel="description"

Reviewer 3: Should `rel="description"` be `rel="describes"`?

along with:

Link: <https://example.com/binary>; rel="describedby"

> For cases where link relations are not defined by IANA, a URL can be used. For example:

Reviewer 3: I think the first sentence would suffice for this block by just referring to RFC Web Linking ( https://tools.ietf.org/html/rfc8288#section-2.1 ) on allowed relation type ie. registered terms at IANA or URI. This information is applicable to both Metadata Discovery and Discovery of Annotated Resource sections. I think it should come before those two sections... either a unique section describing common requirements or maybe even move it to Overview.

If you really want URL, as opposed to a URI, make sure that's well-defined somewhere. On a related note, there is discussion in solid/specification about dereferenceable and machine-readable constraints (eg. target of `ldp:constrainedBy`). I would put these things in the same category even though the former is about the vocab and the latter is some resource.

Author: Per suggestion, pointed to RDF8288 section on allowed types, and removed the latter section that described what is covered in the linked reference.

Link: <https://example.com/resource?meta>;
        rel="https://example.org/ns#hasMetadata"

and:

Link: <https://example.com/resource>;
        rel="https://example.org/ns#isMetadataOf"

String-based link relations, such as in the examples above, must be registered with IANA. But it is also possible to use custom relation types by using a full IRI. The [Linked Data Platform](https://www.w3.org/TR/ldp/), the [Linked Data Notifications](https://www.w3.org/TR/ldn/), and the [Web Annotation Protocol](http://www.w3.org/TR/annotation-protocol/) specifications make use of full IRIs in the `rel` attribute.

## Metadata Characteristics

> A given resource MAY Link to metadata on a different server.

Reviewer 3: How? Just by client referring to it when the resource is created? Or is this something only the server may establish?

If and how can the URI be updated?

Author: In the current definition this is something only the server may establish. The purpose of this was to point out that a server could be configured to link to metadata resources somewhere else.

Reviewer 3: I'm not sure if there is anything prescribed here that can be tested for required interop besides that a server should be capable of including Link headers for metadata relations for a resource and that the target of the relation can be under any authority. Everything else seems to be out of scope.

The prescription for server's capability can actually be derived from discovery mechanism: https://github.com/solid/data-interoperability-panel/pull/32#issuecomment-578549345

Would this clarify:

"Servers that want to associate a metadata resource to a resource MUST either use the Link header with a rel value of a metadata type or asset the relation in the encoded RDF graph of resource's representation."

> Metadata resources on the same Solid server MUST adhere to the same interaction model used by standard Solid resources.

Reviewer 3: This is a bit overloaded I think.. or just unclear. Is it just that metadata resources are like every other resource? Is the key point about the resource lifecycle?

Author: Yes - the point is to say that if this resource is a solid resource, it abides by the same interaction model as any other solid resource. We say "the same solid server" because the server may not know that the link to a resource somewhere else is a resource on a solid server.

> ## Reserved Metadata Types

Reviewer 3: What's the actual process to refer and create resource metadata?

Are resource metadata listed in containment? https://github.com/solid/specification/issues/116

Author: The server through configuration determines where metadata resources live and provides that in link headers. Assuming a given metadata resource is on the same solid server, it should be listed in containment in that location, and follow the same interaction model as other solid resources, with the exception of any special handling due a resource of a given metadata type.

### Web Access Control

ACL resources as defined by [Web Access Control](https://github.com/solid/web-access-control-spec) MUST be supported as a resource metadata type by Solid servers.

ACL metadata resources are discoverable by the client via ```rel=acl```.

A given Solid resource MUST NOT be directly associated with more than one ACL metadata resource.

> An ACL metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.

Reviewer 3: Move delete information that applies to all categories to parent section.

> An ACL metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
An ACL metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

> To access or manage an ACL metadata resource, an [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) MUST have [acl:Control](https://github.com/solid/web-access-control-spec#aclcontrol) privileges per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

Reviewer 3: I don't think this information needs to be said (and may even go stale if changed at source; WAC). Perhaps something like this (which similar to the ones in the other metadata categories) move into an earlier section about all categories. Like the example with deleting a resource also deletes its metadata?

Author: We say now in the Metadata Characteristics section that permissions will be defined in the definition of a given type. The ACL permissions differ by metadata type, so it wouldn't be appropriate to say more than that. The delete behavior is consistent across the currently defined types. A new type could have reason to be handled differently, but delete would be okay to move into a general section given the current set. We decided early on that we'd overload the individual types and put less in the general description, which is why we include it in each one.

> A Solid server SHOULD sanity check ACL metadata resources upon creation or update to restrict invalid changes.

Reviewer 3: Nice but perhaps underspecified. Is this supposed to be about having a shape for ACL? Are there cases where a server may let an invalid change stay?

Is this actually about resource metadata or resources in general? If latter, it should refer to some common behaviour.

Reviewer: @ericprud made PR with shapes for ACL resources https://github.com/solid/web-access-control-spec/pull/77/files I think we could reference them instead of talking in terms of "sanity check".

Author: I think shape is part of the solution, but not all. There is some permission logic that a shape can't cover on its own.

### Resource Description

Resource description is a general mechanism to provide descriptive metadata for a given resource. It MUST be supported as a resource metadata type by Solid servers.

> The Descriptive metadata resource for a given resource is discovered by the client via ```rel=describedby```. Conversely, the resource being described by a Descriptive metadata resource is discovered by the client via ```rel=http://www.w3.org/ns/solid/terms#resource```.

Reviewer 3: What's `solid:resource` supposed to be? Aside: I don't like the name. What happened to `describes`?

A given Solid resource MUST NOT be directly associated with more than one Descriptive metadata resource.

> A Descriptive metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
A Descriptive metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

Access or management of a Descriptive metadata resource by a given [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) is subject to the [modes of access](https://github.com/solid/web-access-control-spec#modes-of-access) granted per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

> ### Shape Validation

Reviewer 3: [What shapes the shapes?  Okay, I'll see myself out.]

Shape Validation ensures that any data changes in a given resource conform to an associated [SHACL](https://www.w3.org/TR/shacl/) or [ShEx](https://shex.io/shex-semantics/index.html) data shape. It MUST be supported as a resource metadata type by Solid servers.

The Shape validation metadata resource for a given resource is discovered by the client via ```rel=http://www.w3.org/ns/solid/terms#shape```. Conversely, the resource being described by a Shape validation metadata resource is discovered by the client via ```rel=http://www.w3.org/ns/solid/terms#resource```.

A given Solid resource MUST NOT be directly associated with more than one Descriptive metadata resource.

> A Shape validation metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
A Shape validation metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

To access or manage a Shape validation metadata resource, an [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) MUST have [acl:Control](https://github.com/solid/web-access-control-spec#aclcontrol) privileges per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

A Solid server SHOULD sanity check Shape validation metadata resources upon creation or update to restrict invalid changes.

### Server Managed

> A Solid server stores information about a resource that clients can read but not change in Server Managed metadata. It MUST be supported as a resource metadata type by Solid servers.

Reviewer 3: Perhaps give more detail about "server managed" or tie it to use cases.

> A Server Managed metadata resource is discovered by the client via ```rel=http://www.w3.org/ns/solid/terms#servermanaged```. Conversely, the resource being described by a Server Managed metadata resource is discovered by the client via ```rel=http://www.w3.org/ns/solid/terms#resource```.

Reviewer 3: Minor: Can we use a one word term instead of "servermanaged"?

Author: Suggestions welcome!! 😄

Author: shortened the term to ```rel=http://www.w3.org/ns/solid/terms#managed``` 😐 😄

A given Solid resource MUST NOT be directly associated with more than one Server Managed metadata resource.

> A Server Managed metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
A Server Managed metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

To access a Server Managed metadata resource, an [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) MUST have [acl:Read](https://github.com/solid/web-access-control-spec#aclread) privileges per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

### Configuration

> Configuration metadata is used to store configurable parameters for a given resource. It MUST be supported as a resource metadata type by Solid servers.

Reviewer 3: Perhaps give more detail about "configuration" or tie it to use cases.

A configuration metadata resource is discovered by the client via ```rel=http://www.w3.org/ns/solid/terms#configuration```. Conversely, the resource being described by a Configuration metadata resource is discovered by the client via ```rel=http://www.w3.org/ns/solid/terms#resource```.

A given Solid resource MUST NOT be directly associated with more than one Configuration metadata resource.

> A Configuration metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
A Configuration metadata resource MUST be deleted by the Solid server when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

To access or manage a Configuration metadata resource, an [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) MUST have [acl:Control](https://github.com/solid/web-access-control-spec#aclcontrol) privileges per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

> ## Non-Reserved Types

Reviewer 3: I'd suggest to express this in terms of "Extensions" and mention that it needs to follow the base or global criteria as the other metadata types.

A Solid server may support other metadata types.

> ## Implementation Patterns

Reviewer 3: I don't know if you plan to extend this section but may want to tuck it under a section about common behaviour early on.

For cases where link relations are not defined by IANA, a URL can be used. For example:

Link: <https://example.com/resource?meta>;
        rel="https://example.org/ns#hasMetadata"

and:

Link: <https://example.com/resource>;
        rel="https://example.org/ns#isMetadataOf"

String-based link relations, such as in the examples above, must be registered with IANA. But it is also possible to use custom relation types by using a full IRI. The [Linked Data Platform](https://www.w3.org/TR/ldp/), the [Linked Data Notifications](https://www.w3.org/TR/ldn/), and the [Web Annotation Protocol](http://www.w3.org/TR/annotation-protocol/) specifications make use of full IRIs in the `rel` attribute.

## Metadata Characteristics

A given resource MAY Link to metadata on a different server.

Metadata resources on the same Solid server MUST adhere to the same interaction model used by standard Solid resources.

## Reserved Metadata Types

### Web Access Control

ACL resources as defined by [Web Access Control](https://github.com/solid/web-access-control-spec) MUST be supported as a resource metadata type by Solid servers.

ACL metadata resources are discoverable by the client via ```rel=acl```.

> An ACL metadata resource MUST be deleted when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
An ACL metadata resource MUST be deleted when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

To access or manage an ACL meta resource, an [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) MUST have [acl:Control](https://github.com/solid/web-access-control-spec#aclcontrol) privileges per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

A Solid server SHOULD sanity check ACL metadata resources upon creation or update to restrict invalid changes.

### Resource Description

Resource description is a general mechanism to provide descriptive metadata for a given resource. It MUST be supported as a resource metadata type by Solid servers.

The Descriptive metadata resource for a given resource is discovered via ```rel=describedby```. Conversely, the resource being described by a Descriptive metadata resource is discovered via ```rel=http://www.w3.org/ns/solid/terms#resource```.

> A Descriptive metadata resource MUST be deleted when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
A Descriptive metadata resource MUST be deleted when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

Access or management of a Descriptive metadata resource by a given [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) is subject to the [modes of access](https://github.com/solid/web-access-control-spec#modes-of-access) granted per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

### Shape Validation

Shape Validation ensures that any data changes in a given resource conform to an associated [SHACL](https://www.w3.org/TR/shacl/) or [ShEx](https://shex.io/shex-semantics/index.html) data shape. It MUST be supported as a resource metadata type by Solid servers.

The Shape validation metadata resource for a given resource is discovered via ```rel=http://www.w3.org/ns/solid/terms#shape```. Conversely, the resource being described by a Shape validation metadata resource is discovered via ```rel=http://www.w3.org/ns/solid/terms#resource```.

> A Shape validation metadata resource MUST be deleted when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
A Shape validation metadata resource MUST be deleted when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

To access or manage a Shape validation metadata resource, an [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) MUST have [acl:Control](https://github.com/solid/web-access-control-spec#aclcontrol) privileges per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

A Solid server SHOULD sanity check Shape validation metadata resources upon creation or update to restrict invalid changes.

### Server Managed

A Solid server stores information about a resource that clients can read but not change in Server Managed metadata. It MUST be supported as a resource metadata type by Solid servers.

A Server Managed metadata resource is Discovered via ```rel=http://www.w3.org/ns/solid/terms#server```. Conversely, the resource being described by a Server Managed metadata resource is discovered via ```rel=http://www.w3.org/ns/solid/terms#resource```.

> A Server Managed metadata resource MUST be deleted when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
A Server Managed metadata resource MUST be deleted when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

To access a Server Managed metadata resource, an [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) MUST have [acl:Read](https://github.com/solid/web-access-control-spec#aclread) privileges per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

### Configuration

Configuration metadata is used to store configurable parameters for a given resource. It MUST be supported as a resource metadata type by Solid servers.

A configuration metadata resource is discovered via ```rel=http://www.w3.org/ns/solid/terms#configuration```. Conversely, the resource being described by a Configuration metadata resource is discovered via ```rel=http://www.w3.org/ns/solid/terms#resource```.

> A Configuration metadata resource MUST be deleted when the resource it is directly associated with is also deleted and the Solid server is authoritative for both resources.

Reviewer 2: ```suggestion
A Configuration metadata resource MUST be deleted when the resource it is directly associated with is deleted and the Solid server is authoritative for both resources.
```

A given resource might link to zero or more such related metadata representations. Having a unified model to describe how clients and servers interact with and process these resources will help clarify expectations, while also providing a shared pattern for extending the features of server implementations.

The metadata model described in this document makes it possible for clients to decorate resources with structured descriptive data. Those metadata may serve as supplementary descriptions or, when supported by a server implementation, they may influence the behavior of the resources.

Some examples include:

- An ACL resource controls how a server makes authorization decisions for a container and any child resources.
- A shape constraint resource may limit the RDF structures that can be added to a resource or container.
- A “configuration” resource may control how a resource is versioned, or which indexes are exposed for it.

Access to different categories of metadata may require different levels of privilege, which must be specified as part of the definition for a given metadata type.

## Metadata Discovery

> Given the URL of a resource, a client can discover the metadata resources by issuing a GET (or HEAD) request and inspecting the Link headers in the response. The [rel= attribute](https://tools.ietf.org/html/rfc8288) will define the relationship to the target URL.

Reviewer 2: ```suggestion
Given the URL of a resource, a client can discover the metadata resources by issuing a GET (or HEAD) request and inspecting the Link headers in the response. The [`rel={attribute}`](https://tools.ietf.org/html/rfc8288) will define the relationship to the target URL.
```

A Descriptive metadata resource MUST be deleted when the resource it is directly associated with is also deleted and the two resources reside on the same Solid server.

Access or management of a Descriptive metadata resource by a given [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) is subject to the [modes of access](https://github.com/solid/web-access-control-spec#modes-of-access) granted per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

### Shape Validation

Shape Validation ensures that any data changes in a given resource conform to an associated [SHACL](https://www.w3.org/TR/shacl/) or [ShEx](https://shex.io/shex-semantics/index.html) data shape. It MUST be supported as a resource metadata type by Solid servers.

The Shape validation metadata resource for a given resource is discovered via ```rel=http://www.w3.org/ns/solid/terms#shape``` and ```rel=http://www.w3.org/ns/solid/terms#resource```.

> .```OPEN QUESTION - Should the link be directly to the shape resource (which is often not going to be on the same server)? Should it be to a local resource that in turn points to the shape (wherever it is), but can also include configuration parameters on the nature of the validation? The latter would seem to provide much more flexibility in the long run.```

Reviewer: As I understand, server has full and exclusive control when it comes to creating all the 'metadata resources' and links between them and the resource. Client would not have any direct control over those link relations and could only have to those metadata resources. In that case for client to have possibility to reference some specific shape that resource should validate against, it would need to modify specific metadata resource linked by the server with `solid:shape` relation.

Author: Resolve in most recent version.

A Shape validation metadata resource MUST be deleted when the resource it is directly associated with is also deleted and the two resources reside on the same Solid server. ```Note: Assumes that we follow the local resource pointing to shape approach```

To access or manage a Shape validation metadata resource, an [acl:agent](https://github.com/solid/web-access-control-spec#describing-agents) MUST have [acl:Control](https://github.com/solid/web-access-control-spec#aclcontrol) privileges per the [ACL inheritance algorithm](https://github.com/solid/web-access-control-spec#acl-inheritance-algorithm) on the resource directly associated with it.

A Solid server SHOULD sanity check Shape validation metadata resources upon creation or update to restrict invalid changes.

### Server Managed

*This section is still in process*

Server-managed metadata not otherwise included in HTTP headers. MUST be supported.

> Discovered via ```rel=http://www.w3.org/ns/solid/terms#server``` and ```rel=http://www.w3.org/ns/solid/terms#resource```

Reviewer: If server doesn't advertise them in HTTP Link Header where would those link relations appear and who would try to perform discovery based on them?
